src/progress.rs

```diff
@@ -47,119 +47,190 @@ impl Default for ProgressState {
+        let ids = &self.configuration.voters;
+        self.progress
+            .iter_mut()
+            .filter(move |(k, _)| ids.contains(k))
+    }
+
+    /// Returns the mutable status of learners.
+    #[inline]
+    pub fn learners_mut(&mut self) -> impl Iterator<Item = (&u64, &mut Progress)> {
+        let ids = &self.configuration.learners;
+        self.progress
+            .iter_mut()
+            .filter(move |(k, _)| ids.contains(k))
+    }
+
+    /// Returns the ids of all known voters.
+    #[inline]
+    pub fn voter_ids(&self) -> &FxHashSet<u64> {
+        &self.configuration.voters
     }
 
     /// Returns the ids of all known learners.
-    pub fn learner_nodes(&self) -> Vec<u64> {
-        let mut ids = Vec::with_capacity(self.learners.len());
-        ids.extend(self.learners.keys());
-        ids.sort();
-        ids
+    #[inline]
+    pub fn learner_ids(&self) -> &FxHashSet<u64> {
+        &self.configuration.learners
     }
 
     /// Grabs a reference to the progress of a node.
+    #[inline]
     pub fn get(&self, id: u64) -> Option<&Progress> {
-        self.voters.get(&id).or_else(|| self.learners.get(&id))
+        self.progress.get(&id)
     }
 
     /// Grabs a mutable reference to the progress of a node.
+    #[inline]
     pub fn get_mut(&mut self, id: u64) -> Option<&mut Progress> {
-        let progress = self.voters.get_mut(&id);
-        if progress.is_none() {
-            return self.learners.get_mut(&id);
-        }
-        progress
+        self.progress.get_mut(&id)
     }
 
     /// Returns an iterator across all the nodes and their progress.
-    pub fn iter(&self) -> impl Iterator<Item = (&u64, &Progress)> {
-        self.voters.iter().chain(&self.learners)
+    #[inline]
+    pub fn iter(&self) -> impl ExactSizeIterator<Item = (&u64, &Progress)> {
+        self.progress.iter()
     }
 
     /// Returns a mutable iterator across all the nodes and their progress.
-    pub fn iter_mut(&mut self) -> impl Iterator<Item = (&u64, &mut Progress)> {
-        self.voters.iter_mut().chain(&mut self.learners)
+    #[inline]
+    pub fn iter_mut(&mut self) -> impl ExactSizeIterator<Item = (&u64, &mut Progress)> {
+        self.progress.iter_mut()
     }
 
     /// Adds a voter node
-    pub fn insert_voter(&mut self, id: u64, pr: Progress) -> Result<(), Error> {
-        if self.voters.contains_key(&id) {
-            Err(Error::Exists(id, "voters"))?
-        }
-        if self.learners.contains_key(&id) {
-            Err(Error::Exists(id, "learners"))?;
+    pub fn insert_voter(&mut self, id: u64, mut pr: Progress) -> Result<(), Error> {
+        // If the progress exists already this is in error.
+        if self.progress.contains_key(&id) {
+            // Determine the correct error to return.
+            if self.learner_ids().contains(&id) {
+                return Err(Error::Exists(id, "learners"));
+            } else if self.voter_ids().contains(&id) {
```

Comment:
The check seems redundant.

Reply:
I do not think inserting a voter who is already a voter should be silently ignored. It is a clear sign that the user thinks the system is in a different state than it is. Performance wise this is completely non-consequential.

Reply:
What I mean if it exists in `progress` and not in `learner_ids`, then it must be in `voter_ids`.